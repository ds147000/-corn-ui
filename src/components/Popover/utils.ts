/* eslint-disable no-magic-numbers */
/* eslint-disable no-unreachable */
// #if _APP === 'weapp'
import Taro from '@tarojs/taro'
// #endif

export interface ElementRect {
  width: number
  height: number
  top: number
  bottom: number
  left: number
  right: number
  offsetLeft: number
  offsetTop: number
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export async function getElementRect(el: any): Promise<ElementRect> {
  const result: ElementRect = {
    width: 0,
    height: 0,
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    offsetLeft: 0,
    offsetTop: 0
  }
  // #if _APP !== 'weapp'
  const _el = el as HTMLDivElement
  const clientRect = _el.getBoundingClientRect()
  result.bottom = clientRect.bottom
  result.top = clientRect.top
  result.left = clientRect.left
  result.right = clientRect.right
  result.height = clientRect.height
  result.width = clientRect.width
  result.offsetLeft = _el.offsetLeft
  result.offsetTop = _el.offsetTop
  return result

  // #else
  return new Promise((res) => {
    Taro.createSelectorQuery()
      .select(`#${el.uid}`)
      .fields({ size: true, properties: [ 'offsetLeft', 'offsetTop' ] }, (response) => {
        res(response as ElementRect)
      })
  })
  // #endif
}
