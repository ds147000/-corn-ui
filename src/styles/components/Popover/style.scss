.xrk-popover {
  display: inline-block;

  &-plane {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    z-index: $portalIndex;
  }

  &-show {
    animation: XrkPopoverShow 0.2s;
  }

  &-hide {
    animation: XrkPopoverHide 0.2s;
  }
}

@keyframes XrkPopoverShow {
  0% {
    transform: scale(0.7);
    opacity: 0.7;
    transform-origin: center;
  }

  100% {
    transform: scale(1);
    opacity: 1;
    transform-origin: center;
  }
}

@keyframes XrkPopoverHide {
  0% {
    transform: scale(1);
    opacity: 1;
    transform-origin: center;
  }

  100% {
    transform: scale(0.3);
    opacity: 0;
    transform-origin: center;
  }
}
