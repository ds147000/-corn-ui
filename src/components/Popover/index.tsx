import React, { useCallback, useEffect, useRef, useState } from 'react'
import { View } from '@tarojs/components'
import { ElementRect, getElementRect } from './utils'

export type PopoverRender = (rect: ElementRect) => JSX.Element

export interface PopoverProps {
  content: PopoverRender
  className?: string
  style?: React.CSSProperties
}

const Popover: React.FC<PopoverProps> = ({ children, className, style, content }) => {
  const [ visibility, setVisibility ] = useState(false)
  const [ isHide, setIsHide ] = useState(false)
  const [ rect, setRect ] = useState<ElementRect>({
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    height: 0,
    width: 0,
    offsetLeft: 0,
    offsetTop: 0
  })
  const El = useRef()

  const onChange = async (): Promise<void> => {
    if (visibility === false) {
      const _rect = await getElementRect(El.current)
      setRect(_rect)
    }

    setVisibility(!visibility)
  }

  const onClose = useCallback(() => setVisibility(false), [])

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const onAnimatedEnd = useCallback((event: any) => {
    const animationName: string = event.animationName ||
      event.detail?.animationName ||
      event.nativeEvent?.animationName ||
      event.target.className

    if (animationName.toLocaleUpperCase().indexOf('HIDE') !== -1) setIsHide(false)
  }, [])

  useEffect(() => {
    if (visibility) setIsHide(true)
  }, [ visibility ])

  return (
    <>
      {isHide && (
        <>
          <View
            className="corn-popover-plane"
            onClick={onClose}
            onTouchStart={onClose}
            data-testid="mask"
          />
          <View
            className={visibility ? 'corn-popover-show' : 'corn-popover-hide'}
            onAnimationEnd={onAnimatedEnd}
            onClick={onClose}
            data-testid="centent"
          >
            {content?.(rect)}
          </View>
        </>
      )}
      <View className={`corn-popover ${className || ''}`} style={style} onClick={onChange} ref={El} >
        {children}
      </View>
    </>
  )
}

export default Popover
