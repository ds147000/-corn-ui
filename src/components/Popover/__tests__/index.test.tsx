import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import Popover from '../index'

describe('Popover', () => {
  test('基础快照', async () => {
    const screen = render(
      <Popover content={() => (
        <div>气泡标题</div>
      )} >
        <div>气泡按钮</div>
      </Popover>
    )

    fireEvent.click(screen.getByText('气泡按钮'))

    await waitFor(() => expect(screen.queryByText('气泡标题')).not.toBeNull())
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('打开', async () => {
    const screen = render(
      <Popover content={() => (
        <div>气泡标题</div>
      )} >
        <div>气泡按钮</div>
      </Popover>
    )

    fireEvent.click(screen.getByText('气泡按钮'))
    await waitFor(() => screen.getByText('气泡标题'))
  })

  test('打开 动画结束', async () => {
    const screen = render(
      <Popover content={() => (
        <div>气泡标题</div>
      )} >
        <div>气泡按钮</div>
      </Popover>
    )

    fireEvent.click(screen.getByText('气泡按钮'))
    await waitFor(() => screen.getByText('气泡标题'))
    fireEvent.animationEnd(screen.getByTestId('centent'), { detail: { animationName: 'SHOW' } })
    await waitFor(() => screen.getByText('气泡标题'))
  })

  test('收起', async () => {
    const screen = render(
      <Popover content={() => (
        <div>气泡标题</div>
      )} >
        <div>气泡按钮</div>
      </Popover>
    )

    fireEvent.click(screen.getByText('气泡按钮'))
    await waitFor(() => screen.getByText('气泡标题'))
    fireEvent.click(screen.getByText('气泡按钮'))
    fireEvent.animationEnd(screen.getByTestId('centent'), { detail: { animationName: 'HIDE' } })
    await waitFor(() => expect(screen.queryByText('气泡标题')).toBeNull())
  })

  test('触摸内容', async () => {
    const screen = render(
      <Popover content={() => (
        <div>气泡标题</div>
      )} >
        <div>气泡按钮</div>
      </Popover>
    )

    fireEvent.click(screen.getByText('气泡按钮'))
    await waitFor(() => screen.getByText('气泡标题'))
    fireEvent.touchStart(screen.getByText('气泡标题'))
    fireEvent.animationEnd(screen.getByTestId('centent'), { detail: { animationName: 'HIDE' } })
    await waitFor(() => expect(screen.queryByText('气泡标题')).not.toBeNull())
  })



  test('面板收起', async () => {
    const screen = render(
      <Popover content={() => (
        <div>气泡标题</div>
      )} >
        <div>气泡按钮</div>
      </Popover>
    )

    fireEvent.click(screen.getByText('气泡按钮'))
    await waitFor(() => screen.getByText('气泡标题'))
    fireEvent.click(screen.getByText('气泡标题'))
    fireEvent.animationEnd(screen.container.querySelector('.corn-popover-hide'), { detail: { animationName: 'HIDE' } })
    await waitFor(() => expect(screen.queryByText('气泡标题')).toBeNull())
  })

  test('设置样式', () => {
    const screen = render(
      <Popover className="abc" content={() => (
        <div>气泡标题</div>
      )} >
        <div>气泡按钮</div>
      </Popover>
    )

    expect(screen.container.getElementsByClassName('abc')).not.toBeNull()
  })
})
