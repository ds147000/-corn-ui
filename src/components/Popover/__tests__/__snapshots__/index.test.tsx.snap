// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover 基础快照 1`] = `
<div>
  <div
    class="xrk-popover-plane"
    data-testid="mask"
  >
    <div
      class="xrk-popover-show"
      data-testid="centent"
    >
      <div>
        气泡标题
      </div>
    </div>
  </div>
  <div
    class="xrk-popover "
  >
    <div>
      气泡按钮
    </div>
  </div>
</div>
`;
